hazelcast/src/main/java/com/hazelcast/internal/util/JavaVersion.java

```diff
@@ -51,101 +47,115 @@ public enum JavaVersion implements JavaMajorVersion {
-     *
-     * @param version version to be compared against the current runtime version
-     * @return Return true if current runtime version of Java is the same or less than given version.
+     * Check if the current runtime version is less than or equal to the given version.
+     * Returns false if the given version is UNKNOWN.
      */
     public static boolean isAtMost(JavaVersion version) {
         return isAtMost(CURRENT_VERSION, version);
     }
 
     static boolean isAtLeast(JavaMajorVersion currentVersion, JavaMajorVersion minVersion) {
-        return currentVersion.getMajorVersion() >= minVersion.getMajorVersion();
+        return currentVersion != UNKNOWN_VERSION && minVersion != UNKNOWN_VERSION
+                && currentVersion.getMajorVersion() >= minVersion.getMajorVersion();
     }
 
     static boolean isAtMost(JavaMajorVersion currentVersion, JavaMajorVersion maxVersion) {
-        return currentVersion.getMajorVersion() <= maxVersion.getMajorVersion();
+        return currentVersion != UNKNOWN_VERSION && maxVersion != UNKNOWN_VERSION
+                && currentVersion.getMajorVersion() <= maxVersion.getMajorVersion();
     }
 
-    private static JavaVersion valueOf(int majorVersion) {
-        for (JavaVersion version : values()) {
-            if (version.majorVersion == majorVersion) {
-                return version;
-            }
+    /**
+     * <pre>
+     * <= JDK 8
+     *     VNUM  := 1\.[1-9]\.[0-9](_[0-9]{2})?         Version number
+     *     PRE   := [a-zA-Z0-9]+                        Pre-release identifier
+     *     java.version
+     *           := $VNUM(-$PRE)*
+     * >= JDK 9
+     *     VNUM  := [1-9][0-9]*(\.(0|[1-9][0-9]*))*     Version number
+     *     PRE   := [a-zA-Z0-9]+                        Pre-release identifier
+     *     BUILD := 0|[1-9][0-9]*                       Build number
+     *     OPT   := [-a-zA-Z0-9\.]+                     Build information
+     *     java.version
+     *           := $VNUM(-$PRE)?\+$BUILD(-$OPT)?
+     *            | $VNUM-$PRE(-$OPT)?
+     *            | $VNUM(+-$OPT)?
+     * </pre>
+     * @see <a href="https://www.oracle.com/java/technologies/javase/versioning-naming.html">
+     *      J2SE SDK/JRE Version String Naming Convention (<= JDK 8)
+     * @see <a href="https://openjdk.org/jeps/223">
+     *      JEP 223: New Version-String Scheme (>= JDK 9)
+     */
+    static JavaMajorVersion detectCurrentVersion() {
+        ILogger logger = Logger.getLogger(JavaVersion.class);
+        String version = System.getProperty("java.version");
+        int major;
+        try {
+            String[] components = version.split("[-+.]");
+            major = Integer.parseInt(components[components[0].equals("1") ? 1 : 0]);
+        } catch (NullPointerException e) {
+            warn(logger, "java.version property is not set");
+            return UNKNOWN_VERSION;
+        } catch (ArrayIndexOutOfBoundsException | NumberFormatException e) {
```

Comment:
Cleaner way would be a null check and array length check than catching exceptions afterward.

Reply:
Checking the array length instead of letting an exception to be thrown makes the code less readable.